src/gridfs/file.rs

```diff
@@ -0,0 +1,641 @@
+        File::with_gfs_file(gfs, GfsFile::new(id), mode)
+    }
+
+    /// A new file stream with a name-and-id-referenced GridFS file.
+    pub fn with_name(gfs: Store, name: String, id: oid::ObjectId, mode: Mode) -> File {
+        File::with_gfs_file(gfs, GfsFile::with_name(name, id), mode)
+    }
+
+    /// A new file stream from a read file document.
+    pub fn with_doc(gfs: Store, doc: bson::Document) -> File {
+        File::with_gfs_file(gfs, GfsFile::with_doc(doc), Mode::Reading)
+    }
+
+    // Generic new file stream.
+    fn with_gfs_file(gfs: Store, file: GfsFile, mode: Mode) -> File {
+        File {
+            mutex: Arc::new(Mutex::new(())),
+            condvar: Arc::new(Condvar::new()),
+            mode: mode,
+            gfs: gfs,
+            chunk: 0,
+            offset: 0,
+            wpending: Arc::new(ATOMIC_ISIZE_INIT),
+            wbuf: Vec::new(),
+            wsum: Md5::new(),
+            rbuf: Vec::new(),
+            rcache: None,
+            doc: file,
+            err: Arc::new(RwLock::new(InnerError { inner: None })),
+        }
+    }
+
+    /// Returns the byte length of the file.
+    pub fn len(&self) -> i64 {
+        self.len
+    }
+
+    /// Retrieves the description of the threaded error, if one occurred.
+    pub fn err_description(&self) -> Result<Option<String>> {
+        let err = try!(self.err.read());
+        let ref inner = err.deref().inner;
+        let description = match inner {
+            &Some(ref err) => Some(err.description().to_owned()),
+            &None => None
+        };
+        Ok(description)
+    }
+
+    /// Ensures the file mode matches the desired mode.
+    pub fn assert_mode(&self, mode: Mode) -> Result<()> {
+        if self.mode != mode {
+            return match self.mode {
+                Mode::Reading => Err(ArgumentError("File is open for reading.".to_owned())),
+                Mode::Writing => Err(ArgumentError("File is open for writing.".to_owned())),
+                Mode::Closed => Err(ArgumentError("File is closed.".to_owned())),
+            }
+        }
+        Ok(())
+    }
+
+    /// Completes writing or reading and closes the file. This will be called when the
+    /// file is dropped, but errors will be ignored. Therefore, this method should
+    /// be called manually.
+    pub fn close(&mut self) -> Result<()> {
+
+        // Flush chunks
+        if self.mode == Mode::Writing {
+            try!(self.flush());
+        }
+
+        let _ = try!(self.mutex.lock());
+
+        // Complete file write
+        if self.mode  == Mode::Writing {
+            if try!(self.err_description()).is_none() {
+                if self.doc.upload_date.is_none() {
+                    self.doc.upload_date = Some(UTC::now());
+                }
+                self.doc.md5 = self.wsum.result_str();
+                try!(self.gfs.files.insert_one(self.doc.to_bson(), None));
+                //self.gfs.chunks.ensure_index_key("files_id", "n");
```

Comment:
Alright, so I don't need to ensure the index keys, then? I assume any other GridFS tool that needs them will create these indexes if they don't already exist on the collections. 

I'm guessing it would still be ideal to have these indexes to speed up the fs.